pkg/driver/node.go

```diff
@@ -31,267 +29,270 @@ import (
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Infof("Staging target path %s does not exist, nothing to unmount", stagingTarget)
+			return &csi.NodeUnstageVolumeResponse{}, nil
+		}
+		return nil, status.Errorf(codes.Internal, "Failed to unmount staging target path %s: %v", stagingTarget, err)
 	}
 
-	if req.GetVolumeCapability() == nil {
-		return nil, status.Errorf(codes.InvalidArgument, common.NoCapabilitiesSupplied, req.GetVolumeId())
+	return &csi.NodeUnstageVolumeResponse{}, nil
+}
+
+func (d *CSIDriver) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+
+	volume_id := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+	volumeCapability := req.GetVolumeCapability()
+
+	if volume_id == "" || targetPath == "" || volumeCapability == nil {
+		log.WithFields(log.Fields{
+			"volume_id":        volume_id,
+			"targetPath":       targetPath,
+			"volumeCapability": volumeCapability,
+		}).Errorf("Invalid arguments")
+		if volume_id == "" {
+			return nil, status.Error(codes.InvalidArgument, common.EmptyVolumeId)
+		}
+		if targetPath == "" {
+			return nil, status.Error(codes.InvalidArgument, common.EmptyTargetPath)
+		}
+		if volumeCapability == nil {
+			return nil, status.Errorf(codes.InvalidArgument, common.NoCapabilitiesSupplied, volume_id)
+		}
 	}
 
-	defer d.releaseVolumeLock(req.GetVolumeId())
-	d.getVolumeLock(req.GetVolumeId())
+	defer d.releaseVolumeLock(volume_id)
+	d.getVolumeLock(volume_id)
 
-	log.Infof("Attempting to publish volume %s", req.GetVolumeId())
+	log.Infof("Attempting to publish volume %s", volume_id)
 
-	var volumeMode, fsType string
+	var volumeContext = req.GetVolumeContext()
+	var readOnly bool = req.GetReadonly()
 	var mountFlags []string
-	fqdn := req.GetVolumeContext()["fqdn"]
-	cap := req.GetVolumeCapability()
-	switch cap.GetAccessType().(type) {
+	var fsType, backingShareName string
+
+	switch volumeCapability.GetAccessType().(type) {
 	case *csi.VolumeCapability_Block:
-		volumeMode = "Block"
+		backingShareName = volumeContext["blockBackingShareName"]
 	case *csi.VolumeCapability_Mount:
-		volumeMode = "Filesystem"
-		fsType = cap.GetMount().FsType
+		backingShareName = volumeContext["mountBackingShareName"]
+		fsType = volumeCapability.GetMount().FsType
 		if fsType == "" {
-			fsType = req.GetVolumeContext()["fsType"]
+			fsType = volumeContext["fsType"]
 			if fsType == "" {
 				fsType = "nfs"
 			}
 		}
-		mountFlags = cap.GetMount().MountFlags
+		mountFlags = volumeCapability.GetMount().MountFlags
 	default:
-		return nil, status.Errorf(codes.InvalidArgument, common.NoCapabilitiesSupplied, req.GetVolumeId())
+		return nil, status.Errorf(codes.InvalidArgument, common.NoCapabilitiesSupplied, volume_id)
 	}
-	var err error
-	if fsType == "nfs" {
-		err = d.publishShareBackedVolume(req.GetVolumeId(), req.GetTargetPath(), mountFlags, req.GetReadonly(), fqdn)
+
+	// For NFS
+	if fsType == "nfs" && backingShareName == "" {
```

Comment:
Case when user want to use nfs volumes without any base backing share
Same as we are running currently.

In case of dynamic pvc it will created a new share.
Will take some time this share to come up, but once it up, it be bind attached to that share.

If pvc is already bounded to pv then we will be already have that volume in base share / and the path will match and a bind mount is created.